### PR TITLE
Add CLI flag for aggregation timeout

### LINKED ISSUE #482: Make sonobuoy timeout configurable from the CLI
**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Add a command-line option to make the time out configurable to `sonobuoy run` and `sonobuoy gen`.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

I'm using sonobuoy for to run the conformance tests in an environment where it's typical for it to take between 5 and 6 hours to finish the tests.  A significant chunk of this time is waiting for tests to hit their time outs when they fail, but I also have environments where cluster operations simply take a long time.  In v.0.10, I increased the timeout by creating a private branch of the repo; in v0.11.3 I'm having to rewrite the JSON in the ConfigMap.

**Environment:**

- Sonobuoy version: v0.11.3
- Kubernetes version: (use `kubectl version`): 1.10, OpenShift 3.9
- Kubernetes installer & version: kubespray and OpenShift
- Hardware configuration: VMs or bare metal
- OS (e.g. from `/etc/os-release`):  Ubuntu or RHEL

### PR DESCRIPTION
**Pull request by johnSchnake**
**What this PR does / why we need it**:
One of the most common config settings that may change between
different environments and test cases is the amount of time
Sonobuoy will wait for results. Currently, this value is configurable
but it is in a nested config structure and not terribly well
documented at this time.

Personally, in the past, I had projects where we had to start
using a sonobuoy config-based flow instead of flags due solely
to this field.

This change adds the flag to make it more easy to set.

Signed-off-by: John Schnake <jschnake@vmware.com>

**Which issue(s) this PR fixes**
Fixes #482

**Release note**:
```
Added --timeout flag to set the number of seconds Sonobuoy will wait for plugins to complete.
```

### CODE REVIEW

**Comment by johnSchnake** [comment]
// @ceridwen

**Comment by stevesloka** [comment]
So now there are 2 flags which determine "wait". There's `--wait` and now `--timeout` (added in this PR). Should those do the same thing and be combined?

File: cmd/sonobuoy/app/args.go [hunk 1]
```diff
@@ -253,6 +254,14 @@ func AddRunWaitFlag(flag *int, flags *pflag.FlagSet) {
 	flags.Lookup("wait").NoOptDefVal = "1440"
 }
 
+// AddTimeoutFlag adds an int flag for waiting for the entire run to finish.
+func AddTimeoutFlag(flag *int, flags *pflag.FlagSet) {
+	flags.IntVar(
+		flag, timeoutFlag, config.DefaultAggregationServerTimeoutSeconds,
+		"How long (in seconds) Sonobuoy will wait for plugins to complete before exiting. 0 indicates no timeout.",
```
Comment (stevesloka) [inline cmd/sonobuoy/app/args.go]: Might be cool to make this a duration type and let users enter `1m` or 2h`, etc. Otherwise, you've got to do the math to figure out the number of seconds.

Comment (stevesloka) [reply cmd/sonobuoy/app/args.go]: nit: Maybe print the default to the help so users know what it is if not defined?

**Comment by stevesloka** [comment]
This could be a future design PR, but is there a way to define the timeout per plugin? For instance, the e2e tests may take a long time, but maybe my systemd_logs only takes 2mins? This flag could be the default of all and then later maybe let plugin developers further define?

File: cmd/sonobuoy/app/args.go (hunk 1 above)
Comment (johnSchnake) [reply cmd/sonobuoy/app/args.go]: cobra adds the default; this was copied from my console just now:
```
... before exiting. 0 indicates no timeout. (default 10800)
```

**Comment by johnSchnake** [comment]
> So now there are 2 flags which determine "wait". There's --wait and now --timeout (added in this PR). Should those do the same thing and be combined?

So there are two flags but they really do indicate different things and could be used in various ways.

The `--timeout` flag indicates the server timeout and when that is hit, everything turns off and explodes (more or less).

The `--wait` flag is a CLI-only flag that just indicates how long you are willing to block in the terminal for. It won't err the run just because you hit that limit.

The reason I don't want to sync the two is that a pretty reasonable use case is the following:
 - run and allow the server to timeout after 2hr
 - you really dont think it'll take that long so you just have your CLI stop waiting and return after an hour.
 - after the hour is up you can check the status; you assume it should be done but if not (and it is still running) you can investigate everything while it is still running.

**Comment by johnSchnake** [comment]
> This could be a future design PR, but is there a way to define the timeout per plugin? For instance, the e2e tests may take a long time, but maybe my systemd_logs only takes 2mins? This flag could be the default of all and then later maybe let plugin developers further define?

So I agree something like that may eventually be implemented. There are definitely impacts of this overall timeout and per-plugin timeout that play a role in #627 My POC implementation of plugin ordering also showed me that it could be very easy to adjust some of these details by adding them into the plugin manifest metadata.

But yes, I agree it is for another day since this is currently coded for just one big timeout and this was just to make it easier to set.

**Comment by stevesloka** [comment]
Yeah that holds up, but I can still `sonobuoy run --wait`, then CTRL-C and then `sonobuoy status` and all works. 

> The --wait flag is a CLI-only flag that just indicates how long you are willing to block in the terminal for. It won't err the run just because you hit that limit.

But should it? If you had CI running and this happened you'd get a false positive wouldn't you? 

I'm just trying to challenge the overall design to make sure we're solving the problem and not just throwing flags at it. Is there a scenario where you'd want the client to wait 5 mins and the server wait 2 hours?

**Comment by johnSchnake** [comment]
> Yeah that holds up, but I can still sonobuoy run --wait, then CTRL-C and then sonobuoy status and all works.

True, though you may want to do something like this even in CI. Before the `--wait` commands existed thats what I had seen done when people were rolling their own polling logic. Wait while it runs for a while, then when we hit our own timeout logic go and gather logs and maybe run some checks while it Sonobuoy is still running, then err out ourselves and clean up.

> But should it? If you had CI running and this happened you'd get a false positive wouldn't you?

As the example above says, I don't think forcing the server to error and stop running because we hit a CLI timeout is always the right thing to do. It prevents a user (including CI systems) from investigating if they want to.

Just because `sonobuoy run --wait` returns doesn't guarantee anything about whether or not anything succeeded; do you mean `false positive` in that you might think it is done? That is partially why the default time for `--wait` is much longer than any typical run, currently at a day.

> I'm just trying to challenge the overall design to make sure we're solving the problem and not just throwing flags at it. 

Absolutely 👍 , IMO that is the discussion that we had when deciding to put this ticket into the milestone though. I appreciate the design questions though and am happy to defend the choice to make the new flag.

> Is there a scenario where you'd want the client to wait 5 mins and the server wait 2 hours?

Yes but I also think your comment about CTRL+C plays a role here too; personally, I run checks all the time where this is the [somewhat] the case. If I am iterating on some plugin (in my case, e2e test) and I _think_ it will just take 5m or less and I run it as such. If it runs and I find out it actually took 7m, I _really_ didn't want the server to err at 5m. But the CLI `--wait` gave me an opportunity to see that 5m has passed and it still isn't done. Again, then I investigate the situation and usually restart my experiment whenever I'm done.

TL;DR; I think that in most situations your sentiment is right, you'd just want the server to pass/fail on its own and then the CLI wait for that terminal status. But I can think of situations where you want to ensure the CLI stops and gives you time to investigate before cleaning anything up. I think some of these things are also related to when/how/if plugin resources are cleaned up. In the end though, this is already a customizable value, just one that multiple people edit regularly and want it to be more simple.

If we decide to prioritize per-plugin timeouts, cleanup logic, etc then I think we should bring up these ideas again though because I do acknowledge there are surely some improvements that could be made.
